### PR TITLE
GH-37770: [MATLAB] Add CSV `TableReader` and `TableWriter` MATLAB classes

### LINKED ISSUE #37770: [MATLAB] Add CSV `TableReader` and `TableWriter` MATLAB classes
### Describe the enhancement requested

To enable initial CSV I/O support in the MATLAB interface, we should add a `TableReader` and `TableWriter` MATLAB class which work with `arrow.tabular.Table`s.

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by kevingurney**
### Rationale for this change

To enable initial CSV I/O support, this PR adds `arrow.io.csv.TableReader` and `arrow.io.csv.TableWriter` MATLAB classes to the MATLAB interface.

### What changes are included in this PR?

1. Added a new `arrow.io.csv.TableReader` class
2. Added a new `arrow.io.csv.TableWriter` class

**Example**
```matlab
>> matlabTableWrite = array2table(rand(3))

matlabTableWrite =

  3×3 table

     Var1        Var2       Var3  
    _______    ________    _______

    0.91131    0.091595    0.24594
    0.51315     0.27368    0.62119
    0.42942     0.88665    0.49501

>> arrowTableWrite = arrow.table(matlabTableWrite)

arrowTableWrite = 

Var1: double
Var2: double
Var3: double
----
Var1:
  [
    [
      0.9113083542736461,
      0.5131490075412158,
      0.42942202968065213
    ]
  ]
Var2:
  [
    [
      0.09159480217154525,
      0.27367730380496647,
      0.8866478145458545
    ]
  ]
Var3:
  [
    [
      0.2459443412735529,
      0.6211893868708748,
      0.49500739584280073
    ]
  ]

>> writer = arrow.io.csv.TableWriter("example.csv")

writer = 

  TableWriter with properties:

    Filename: "example.csv"

>> writer.write(arrowTableWrite)

>> reader = arrow.io.csv.TableReader("example.csv")

reader = 

  TableReader with properties:

    Filename: "example.csv"

>> arrowTableRead = reader.read()

arrowTableRead = 

Var1: double
Var2: double
Var3: double
----
Var1:
  [
    [
      0.9113083542736461,
      0.5131490075412158,
      0.42942202968065213
    ]
  ]
Var2:
  [
    [
      0.09159480217154525,
      0.27367730380496647,
      0.8866478145458545
    ]
  ]
Var3:
  [
    [
      0.2459443412735529,
      0.6211893868708748,
      0.49500739584280073
    ]
  ]

>> matlabTableRead = table(arrowTableRead)

matlabTableRead =

  3×3 table

     Var1        Var2       Var3  
    _______    ________    _______

    0.91131    0.091595    0.24594
    0.51315     0.27368    0.62119
    0.42942     0.88665    0.49501

>> isequal(arrowTableRead, arrowTableWrite)

ans =

  logical

   1

>> isequal(matlabTableRead, matlabTableWrite)

ans =

  logical

   1
```

### Are these changes tested?

Yes.

1. Added new CSV I/O tests including `test/arrow/io/csv/tRoundTrip.m` and `test/arrow/io/csv/tError.m`.
2. Both of these test classes inherit from a `CSVTest` superclass.

### Are there any user-facing changes?

Yes.

1. Users can now read and write CSV files using `arrow.io.csv.TableReader` and `arrow.io.csv.TableWriter`.

### Future Directions

1. Expose [options](https://github.com/apache/arrow/blob/main/cpp/src/arrow/csv/options.h) for controlling CSV reading and writing in MATLAB.
2. Add more read/write tests for null value handling and other datatypes beyond numeric and string values.
4. Add a `RecordBatchReader` and `RecordBatchWriter` for CSV.
5. Add support for more I/O formats like Parquet, JSON, ORC, Arrow IPC, etc.

### Notes

1. Thank you @sgilmore10 for your help with this pull request!
2. I chose to add both the `TableReader` and `TableWriter` in one pull request because it simplified testing. My apologies for the slightly lengthy pull request.
* Closes: #37770

### CODE REVIEW

**Review by sgilmore10 — APPROVED** [review]
This looks good! Thanks for adding the base test class and test utilities. We can re-use these in the future for other file types!

**Comment by kevingurney** [comment]
It looks like the `ARROW_CSV` component is currently not enabled in the Arrow C++ libraries `ExternalProject` build. I just pushed a commit that enables this.

**Comment by kevingurney** [comment]
Enabling building of the `ARROW_CSV` component caused some build failures due to `RapidJSON` being inaccessible when building integration tests:

https://github.com/apache/arrow/actions/runs/6225733757/job/16896790846?pr=37773#step:9:681.

We could try to fix these build failures by explicitly installing `RapidJSON` into the GitHub Actions environments. However, considering that we were [already considering removing `GoogleTest` support from the MATLAB build system](#37532) - it might be better to just use this opportunity to do that (this is what the [Python bindings did, as well](https://github.com/apache/arrow/pull/14117)). 

I have already confirmed that not building the Arrow C++ library tests when `ARROW_CSV` is enabled does work on all platforms in another CI job in `mathworks/arrow`:

https://github.com/mathworks/arrow/actions/runs/6226385022

I'll start working on a separate PR to remove `GoogleTest` support from the CMake build system for the MATLAB interface. In the meantime, I will leave this PR open, and I will rebase once `GoogleTest` support has been removed.

**Comment by kevingurney** [comment]
For reference, I am working on the changes required to remove `GoogleTest` support from the CMake build system of the MATLAB interface here: https://github.com/mathworks/arrow/tree/GH-37532.

**Comment by kevingurney** [comment]
For reference - I've opened #37784 for removing `GoogleTest` support.

**Comment by kevingurney** [comment]
**Update**: the code to remove `GoogleTest` support from the MATLAB interface was merged in #37784.

Now that the Arrow C++ tests are no longer being built, enabling `ARROW_CSV` in the MATLAB build works without any issues.

At this point, these changes should be ready to merge.

File: matlab/src/cpp/arrow/matlab/io/csv/proxy/table_reader.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+        mda::ArrayFactory factory;
+
+        // Create a file input stream.
+        MATLAB_ASSIGN_OR_ERROR_WITH_CONTEXT(auto source, arrow::io::ReadableFile::Open(filename, arrow::default_memory_pool()), context, error::FAILED_TO_OPEN_FILE_FOR_READ);
+
+        const ::arrow::io::IOContext io_context;
+        const csv::ReadOptions read_options = csv::ReadOptions::Defaults();
+        const csv::ParseOptions parse_options = csv::ParseOptions::Defaults();
+        const csv::ConvertOptions convert_options = csv::ConvertOptions::Defaults();
```
Comment (kou) [inline matlab/src/cpp/arrow/matlab/io/csv/proxy/table_reader.cc:65]: Can we use `auto` here?

File: matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        mda::StructArray opts = context.inputs[0];
+        const mda::TypedArray<uint64_t> table_proxy_id_mda = opts[0]["TableProxyID"];
+        const uint64_t table_proxy_id = table_proxy_id_mda[0];
+
+        auto proxy = libmexclass::proxy::ProxyManager::getProxy(table_proxy_id);
+        auto table_proxy = std::static_pointer_cast<TableProxy>(proxy);
+        auto table = table_proxy->unwrap();
+
+        MATLAB_ASSIGN_OR_ERROR_WITH_CONTEXT(std::shared_ptr<arrow::io::OutputStream> output_stream,
```
Comment (kou) [inline matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc:77]: Can we use `auto` here?

File: matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                                            arrow::io::FileOutputStream::Open(filename),
+                                            context,
+                                            error::FAILED_TO_OPEN_FILE_FOR_WRITE);
+        csv::WriteOptions options = csv::WriteOptions::Defaults();
```
Comment (kou) [inline matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc:81]: Can we use `auto` here?

**Review by kou — APPROVED** [review]
+1

File: matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc (hunk 2 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc]: Fixed.

File: matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc (hunk 3 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/io/csv/proxy/table_writer.cc]: Fixed.

**Comment by kevingurney** [comment]
The failures for the `Dev / Source Release and Merge Script` CI workflow seem unrelated to this PR.

**Comment by kevingurney** [comment]
CI failures are related to #37803.

**Comment by kevingurney** [comment]
+1

**Comment by kevingurney** [comment]
Unfortunately, the MATLAB CI workflows just started failing due to #37805.

**Comment by kevingurney** [comment]
I've opened #37808 to address the MATLAB CI failures.

I'll merge this PR after #37808 has been merged.

**Comment by kevingurney** [comment]
https://github.com/apache/arrow/pull/37808 has been merged.

I just rebased these changes on top of `main` so that the latest MATLAB CI workflows will run.

**Comment by kevingurney** [comment]
The MATLAB CI workflows are passing now.

I will merge this PR.

**Comment by kevingurney** [comment]
+1

**Pull request merged**